pkg/js/validator/validate-dsl.ts

```diff
@@ -168,166 +172,159 @@ function allowableTypes(typeName: Record<string, TypeDefinition>, type: string,
-// Entry point describes ways that tuples can be assigned to the relation
-// For example,
-// type user
-// type org
-//   relations
-//     define member: [user]
-// we can assign a user with (type user) to org's member
-// However, in the following example
-// type doc
-//   relations
-//     define reader as writer
-//     define writer as reader
-// It is impossible to have any tuples that assign to doc's reader and writer
-function childHasEntryPoint(
-  transformedTypes: Record<string, TypeDefinition>,
-  visitedRecords: Record<string, Record<string, boolean>>,
-  type: string,
-  childDef: RelationTargetParserResult | undefined,
-  allowedTypes: string[],
-): boolean {
-  if (!childDef) {
-    return false;
-  }
-
-  if (childDef.rewrite === RewriteType.Direct) {
-    // we can safely assume that direct rewrite (i.e., it is a self/this), there are direct entry point
-    for (const item of allowedTypes) {
-      const { decodedType, decodedRelation } = destructTupleToUserset(item);
-      if (!decodedRelation) {
-        // this is not a tuple set and is a straight type, we can return true right away
-        return true;
-      }
-      // it is only true if it has unique entry point
-      if (hasEntryPoint(transformedTypes, visitedRecords, decodedType, decodedRelation)) {
-        return true;
-      }
-    }
-  }
-  // otherwise, we will need to follow the child
-  if (!childDef.from) {
-    // this is a simpler case - we only need to check the child type itself
-    if (hasEntryPoint(transformedTypes, visitedRecords, type, childDef.target)) {
-      return true;
-    }
-  } else {
-    // there is a from.  We need to parse thru all the from's possible type
-    // to see if there are unique entry point
-    const fromPossibleTypes = getTypeRestrictions(
-      transformedTypes[type].metadata?.relations![childDef.from].directly_related_user_types || [],
-    );
-
-    for (const fromType of fromPossibleTypes) {
-      const { decodedType } = destructTupleToUserset(fromType);
-
-      // For now, we just look at the type without seeing whether the user set
-      // of the type is reachable too in the case of tuple to user set.
-      // TODO: We may have to investigate whether we need to dive into relation (if present) of the userset
-      if (hasEntryPoint(transformedTypes, visitedRecords, decodedType, childDef.target)) {
-        return true;
-      }
-    }
-  }
-  return false;
-}
-
 // for the type/relation, whether there are any unique entry points
 // if there are unique entry points (i.e., direct relations) then it will return true
 // otherwise, it will follow its children to see if there are unique entry points
 function hasEntryPoint(
   typeMap: Record<string, TypeDefinition>,
+  typeName: string,
+  relationName: string | undefined,
+  rewrite: Userset,
   visitedRecords: Record<string, Record<string, boolean>>,
-  type: string,
-  relation: string | undefined,
-): boolean {
-  if (!relation) {
+): [boolean, boolean] {
```

Comment:
Can we update the comments on what do the two return values mean?  (or update the function name)

Reply:
Adding to the comment, and changed the name to `hasEntryPointOrLoop`